pip/req/req_set.py

```diff
@@ -228,7 +232,11 @@ def prepare_files(self, finder):
             if not self.ignore_installed and not req_to_install.editable:
                 req_to_install.check_if_exists()
                 if req_to_install.satisfied_by:
-                    if self.upgrade:
+                    # check if we are upgrading and that we don't already have
+                    # an exact version match
+                    if self.upgrade and not (
+                            req_to_install.satisfied_by.as_requirement() ==
+                            req_to_install.req):
```

Comment:
Adding the following fixes the problem:

```python
    def __ne__(self, other):
        return not self == other
```

which of course if this is a legitimate bug, should be fixed in setuptools proper and not in the vendored copy of it.

But is this a bug or was there some intentional reason for omitting `__ne__`? @jaraco: Maybe you could shed some light? Let me know if you want me to send a PR to setuptools...